File: src/skaters/periodicity.py
"""Online periodicity detection via running autocorrelation.

Maintains exponentially weighted estimates of autocorrelation at
a set of candidate lags. The top-scoring lags are the detected
periods, which can be used to dynamically add seasonal_difference
transforms during search expansion.

All O(n_lags) per observation. No batch computation.
"""

from __future__ import annotations
import math


# Common periods to scan
DEFAULT_LAGS = [2, 3, 4, 5, 6, 7, 12, 14, 24, 28, 30, 52, 60, 90, 168, 365]


def period_detector(
    lags: list[int] | None = None,
    alpha: float = 0.01,
    min_observations: int = 50,
):
    """Create an online period detector.

    Args:
        lags: candidate periods to test. Defaults to common periods.
        alpha: EMA smoothing for the running autocorrelation estimates.
        min_observations: don't report periods until this many obs seen.

    Returns:
        A callable: (y, state) -> (top_periods, state)
        where top_periods is a list of (lag, acf_score) tuples sorted
        by |acf| descending.
    """
    if lags is None:
        lags = list(DEFAULT_LAGS)

    max_lag = max(lags)

    def _detect(y: float, state: dict | None) -> tuple[list[tuple[int, float]], dict]:
        if state is None:
            state = {
                "buffer": [],
                "n": 0,
                "mean": 0.0,
                "var": 0.0,
                # Per-lag: running EMA of (y_t - mu) * (y_{t-L} - mu)
                "cross": {L: 0.0 for L in lags},
            }

        buf = state["buffer"]
        buf.append(y)
        state["n"] += 1

        # Update running mean and variance
        diff = y - state["mean"]
        state["mean"] += alpha * diff
        state["var"] = (1 - alpha) * (state["var"] + alpha * diff * diff)

        mu = state["mean"]
        var = state["var"]

        # Update cross-correlation for each lag
        for L in lags:
            if len(buf) > L:
                y_lagged = buf[-(L + 1)]
                cross = (y - mu) * (y_lagged - mu)
                state["cross"][L] = (1 - alpha) * state["cross"][L] + alpha * cross

        # Trim buffer
        if len(buf) > max_lag + 1:
            buf.pop(0)

        # Compute ACF scores
        if state["n"] < min_observations or var < 1e-12:
            return [], state

        scores = []
        for L in lags:
            if state["n"] > L:
                acf = state["cross"][L] / var if var > 0 else 0.0
                scores.append((L, acf))

        # Sort by |acf| descending
        scores.sort(key=lambda x: abs(x[1]), reverse=True)
        return scores, state

    return _detect


def top_periods(scores: list[tuple[int, float]], threshold: float = 0.3, max_periods: int = 3) -> list[int]:
    """Extract the best periods from detector scores.

    Args:
        scores: output from period_detector
        threshold: minimum |acf| to consider significant
        max_periods: maximum number of periods to return

    Returns:
        List of detected periods (lags), sorted by strength.
    """
    return [lag for lag, acf in scores[:max_periods] if abs(acf) >= threshold]

File: rust/src/periodicity.rs
//! Online periodicity detection via running autocorrelation. Port of
//! src/skaters/periodicity.py.
//!
//! Maintains exponentially weighted estimates of autocorrelation at a set of
//! candidate lags; the top-scoring lags are the detected periods. O(n_lags)
//! per observation.

use serde::{Deserialize, Serialize};
use std::collections::VecDeque;

/// Common periods to scan.
pub const DEFAULT_LAGS: [usize; 16] = [
    2, 3, 4, 5, 6, 7, 12, 14, 24, 28, 30, 52, 60, 90, 168, 365,
];

/// Online period detector: `step(y)` returns (lag, acf) pairs sorted by
/// |acf| descending (stable sort, so ties keep candidate-lag order, exactly
/// as Python's `list.sort`).
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PeriodDetector {
    pub lags: Vec<usize>,
    pub alpha: f64,
    pub min_observations: u64,
    pub max_lag: usize,
    pub buffer: VecDeque<f64>,
    pub n: u64,
    pub mean: f64,
    pub var: f64,
    /// Per-lag running EMA of (y_t - mu) * (y_{t-L} - mu), parallel to `lags`.
    pub cross: Vec<f64>,
}

impl PeriodDetector {
    /// Python defaults: DEFAULT_LAGS, alpha = 0.01, min_observations = 50.
    pub fn new() -> PeriodDetector {
        PeriodDetector::with_params(None, 0.01, 50)
    }

    pub fn with_params(
        lags: Option<Vec<usize>>,
        alpha: f64,
        min_observations: u64,
    ) -> PeriodDetector {
        let lags = lags.unwrap_or_else(|| DEFAULT_LAGS.to_vec());
        let max_lag = *lags.iter().max().expect("lags must be non-empty");
        let n_lags = lags.len();
        PeriodDetector {
            lags,
            alpha,
            min_observations,
            max_lag,
            buffer: VecDeque::new(),
            n: 0,
            mean: 0.0,
            var: 0.0,
            cross: vec![0.0; n_lags],
        }
    }

    pub fn step(&mut self, y: f64) -> Vec<(usize, f64)> {
        self.buffer.push_back(y);
        self.n += 1;

        // Update running mean and variance
        let diff = y - self.mean;
        self.mean += self.alpha * diff;
        self.var = (1.0 - self.alpha) * (self.var + self.alpha * diff * diff);

        let mu = self.mean;
        let var = self.var;

        // Update cross-correlation for each lag
        let nb = self.buffer.len();
        for (li, &lag) in self.lags.iter().enumerate() {
            if nb > lag {
                let y_lagged = self.buffer[nb - (lag + 1)];
                let cross = (y - mu) * (y_lagged - mu);
                self.cross[li] = (1.0 - self.alpha) * self.cross[li] + self.alpha * cross;
            }
        }

        // Trim buffer
        if self.buffer.len() > self.max_lag + 1 {
            self.buffer.pop_front();
        }

        // Compute ACF scores
        if self.n < self.min_observations || var < 1e-12 {
            return Vec::new();
        }

        let mut scores: Vec<(usize, f64)> = Vec::new();
        for (li, &lag) in self.lags.iter().enumerate() {
            if self.n > lag as u64 {
                let acf = if var > 0.0 { self.cross[li] / var } else { 0.0 };
                scores.push((lag, acf));
            }
        }

        // Sort by |acf| descending (stable; Equal on NaN so a poisoned
        // score cannot panic, matching Python's non-raising sort)
        scores.sort_by(|a, b| {
            b.1.abs()
                .partial_cmp(&a.1.abs())
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        scores
    }
}

impl Default for PeriodDetector {
    fn default() -> PeriodDetector {
        PeriodDetector::new()
    }
}

/// Extract the best periods from detector scores: the first `max_periods`
/// entries with |acf| at or above `threshold`, in strength order.
pub fn top_periods(scores: &[(usize, f64)], threshold: f64, max_periods: usize) -> Vec<usize> {
    scores
        .iter()
        .take(max_periods)
        .filter(|(_, acf)| acf.abs() >= threshold)
        .map(|&(lag, _)| lag)
        .collect()
}
